packages/router/src/grpc/view-protocol-server/nullifier-status.ts

```diff
@@ -1,47 +1,24 @@
+import { MethodImpl } from '@connectrpc/connect';
+import { servicesCtx } from '../../ctx';
+import { ViewProtocolService } from '@buf/penumbra-zone_penumbra.connectrpc_es/penumbra/view/v1alpha1/view_connect';
+import { assertWalletIdMatches } from './utils';
 import {
-  NullifierStatusRequest,
-  NullifierStatusResponse,
   SpendableNoteRecord,
   SwapRecord,
 } from '@buf/penumbra-zone_penumbra.bufbuild_es/penumbra/view/v1alpha1/view_pb';
-import { ViewReqMessage } from './router';
-import { IndexedDbInterface, ServicesInterface } from '@penumbra-zone/types';
-import { assertWalletIdMatches } from './utils';
-import type { Nullifier } from '@buf/penumbra-zone_penumbra.bufbuild_es/penumbra/core/component/sct/v1alpha1/sct_pb';
-
-export const isNullifierStatusRequest = (req: ViewReqMessage): req is NullifierStatusRequest => {
-  return req.getType().typeName === NullifierStatusRequest.typeName;
-};
-
-async function noteOrSwapSpent(
-  indexedDb: IndexedDbInterface,
-  nullifier: Nullifier,
-): Promise<boolean> {
-  const noteByNullifier = await indexedDb.getNoteByNullifier(nullifier);
-  const swapByNullifier = await indexedDb.getSwapByNullifier(nullifier);
 
-  // The 'heightSpent' and 'heightClaimed' fields will never be undefined,
-  // so we compare to 0n assuming it is impossible to spend nullifier in block 0
-  const noteSpent = noteByNullifier ? noteByNullifier.heightSpent !== 0n : false;
-  const swapSpent = swapByNullifier ? swapByNullifier.heightClaimed !== 0n : false;
-  return noteSpent || swapSpent;
-}
-
-export const handleNullifierStatusReq = async (
-  req: NullifierStatusRequest,
-  services: ServicesInterface,
-): Promise<NullifierStatusResponse> => {
+export const nullifierStatus: MethodImpl<
+  typeof ViewProtocolService.methods.nullifierStatus
+> = async (req, ctx) => {
+  const services = ctx.values.get(servicesCtx);
   await assertWalletIdMatches(req.walletId);
   if (!req.nullifier) throw new Error('No nullifier passed');
-
   const { indexedDb } = await services.getWalletServices();
-  const spent = await noteOrSwapSpent(indexedDb, req.nullifier);
+  const swapByNullifier = await indexedDb.getSwapByNullifier(req.nullifier);
+  const noteByNullifier = await indexedDb.getNoteByNullifier(req.nullifier);
+  const spent = Boolean(noteByNullifier?.heightSpent) || Boolean(swapByNullifier?.heightClaimed);
```

Comment:
i think it is pretty widely acknowledged that 0 is falsy

Reply:
what is the appropriate behavior if a spend height of zero does appear on the chain?

Reply:
It's impossible. Block heights start at 1.  Technically speaking, there is not actually a genesis block. There's only a genesis _app state_, which is a state snapshot (rather than a block, which is a kind of state delta). We create a "phantom" compact block with height 0 to allow clients to scan the genesis state snapshot as if it were a delta from the empty state, but it is not a real block.

Reply:
i've added a comment explaining heights. this is one of the handlers that need to be refactored to stream properly so this code will be revisited again pretty soon